[layer] Support MoL backwarding

kparichay opened the pull request:
- Add derivatives for the dot operation as an easier interface to
calculate derivative for a dot operation used in the forward for both
the inputs.
Add the corresponding interface for dot_batched as well.
- Use the dot derivative operation in both fc and attention layer.
- This patch adds bug fix for copy with stride where both the source and
destination tensors are now allowed to be strided tensors.
- Bug fix to work with fc layer with disabled bias.
- Prepare for multiple inheritance for layers by marking public virtual of
the base class.
- Add MoL attention layer backwarding, exportTo, setbatch member
functions.
- Add MoL layer golden test.

See Also #1721
See Also #1614

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

kparichay commented:
Note: MoL backwarding layer has not been optimized to minimize memory usage yet. This implementation focuses on correctness.

zhoonit commented on nntrainer/tensor/tensor.cpp:
```diff
@@ -978,6 +978,68 @@ Tensor Tensor::dot(Tensor const &m, bool trans, bool trans_m) const {
+                                bool trans, bool trans_m, float beta) {
+  bool deriv_trans_m = true;
+  bool deriv_trans = false;
+  /** @todo handle all cases of trans and trans_m */
```
if this is not supported, how about let it throw to prevent unintended use.

I recommend to add tensor test case for all. If this is wrong, we will have many time to figure out.

zhoonit commented on nntrainer/tensor/tensor.cpp:
```diff
@@ -978,6 +978,68 @@ Tensor Tensor::dot(Tensor const &m, bool trans, bool trans_m) const {
+ * @brief compute the derivative of this in the current tensor
+ * @todo will have to see if beta effects this computation
+ */
+Tensor &Tensor::dot_deriv_wrt_1(Tensor const &m, Tensor const &output_deriv,
```
following naming scheme of camel case? (and for others in this commit)

kparichay replied:
Over time, the naming scheme has mixed up camel case (strided, and now this).
I will add an issue to fix the naming scheme for all the cases which is out of scope for this PR.

zhoonit commented on nntrainer/layers/fc_layer.cpp:
```diff
@@ -103,7 +103,7 @@ void FullyConnectedLayer::calcDerivative(RunLayerContext &context) {
   Tensor &ret_ = context.getOutgoingDerivative(SINGLE_INOUT_IDX);
 
-  derivative_.dot(weight, ret_, false, true);
+  ret_.dot_deriv_wrt_1(weight, derivative_, false, false);
```
could you rename `ret_` to `d_input_` or sort for clarity?

kparichay replied:
will do.

zhoonit commented on nntrainer/layers/mol_attention_layer.h:
```diff
@@ -16,14 +16,15 @@
  * @brief   Mixture of Logistics Attention Layer
  */
-class MoLAttentionLayer : public AttentionLayer {
+class MoLAttentionLayer : public AttentionLayer, public LayerImpl {
```
There seems no codebits in this class using AttentionLayer methods, could you explain why trying to enable multiple inheritance?

I think we shouldn't enable multiple inheritance for non-interface classes unless we absolutely need to

kparichay replied:
multiple inheritance will be used once common code bits are removed from mol attention layer code and reused from attention layer.

zhoonit replied:
In that case, we can favor composition, or higher order class. My concern is making LayerImpl virtual, where we are opening a hatch all kinds of multiple inheritance will be possible.

zhoonit replied:
eg1) below will not work without try catch.

```
 AttentionLayer::setProperty(remain_props);
 LayerImpl::setProperty(remain_props);
 exporter.saveResult(mol_props, method, this);
```

eg2) attention layer(or other potential) starts to inherit LayerImpl, this will be wrong

```
 AttentionLayer::exportTo(exporter, method);
 LayerImpl::exportTo(exporter, method);
 exporter.saveResult(mol_props, method, this);
```

kparichay replied:
Agreed, dual inheritance is creating issues with setProperty as pointed in the first example but IMO, LayerImpl should not throw on remaining properties but let the caller throw.
But this is beyond the scope of this PR, so removed inheritance to attention layer.

zhoonit replied:
Thanks :)

kparichay commented:
V2:
[layer] perform layer context check on first forward
Enabled layer context check on the first forward itself,
which revealed a bug in forward which was earlier being shown in the
calcGradient in mol attention layer.
Added the corresponding fix for mol attention layer.
Added corresponding fixes to other locations as well.

zhoonit commented on nntrainer/layers/mol_attention_layer.cpp:
```diff
@@ -72,34 +137,55 @@ void MoLAttentionLayer::forwarding(RunLayerContext &context, bool training) {
+  /** reset helper state */
+  helper_exec = false;
+
+  fc_out = query.dot(fc_w);
```
query.dot(fc_w, fc_out) to maintain reference?

kparichay replied:
this is done in the last commit.

zhoonit commented on nntrainer/layers/mol_attention_layer.cpp:
```diff
@@ -72,34 +137,55 @@ void MoLAttentionLayer::forwarding(RunLayerContext &context, bool training) {
+  tanh.run_fn(fc_out, fc_tanh);
 
-  Tensor mlp_proj_out = mlp_tanh.dot(mlp_proj_w);
+  fc_proj_out = fc_tanh.dot(fc_proj_w);
```
same here?

zhoonit commented on nntrainer/layers/mol_attention_layer.cpp:
```diff
@@ -72,34 +137,55 @@ void MoLAttentionLayer::forwarding(RunLayerContext &context, bool training) {
 
   Tensor alpha;
-  softmax.run_fn(alpha_hat, alpha);
+  softmax.run_fn(alpha_src, alpha);
```
Doesn't this throw considering below?
```
Tensor &ActiFunc::softmax(Tensor const &t, Tensor &output) {
  /**
   * shiftx_logit = logit - max_batch(logit)
   * softmax = exp(shiftx_logit) / (sum(exp(shiftx_logit)))
   *
   * @note softmax is applied on the last dimension
   */
  unsigned int fixed_dim = t.getDim().getDataLen() / t.width();
  float *dp;
  float *rp;

  /** TODO: support strided operations */
  if (t.size() == output.size() && t.getStrides() != output.getStrides())
    throw std::invalid_argument(
      "Softmax does not support operating on strided tensors");      ```

kparichay replied:
because alpha_src is not a shared strided tensor but a contiguous tensor with data copied in it.

kparichay replied:
Let me know if I didnt answer your question.

zhoonit replied:
I wonder alpha.size() is empty?

kparichay replied:
alpha is indeed empty, but it does not throw as empty tensor is not strided

zhoonit replied:
Ah missed that part. Thanks :)

zhoonit commented on nntrainer/layers/mol_attention_layer.cpp:
```diff
@@ -72,34 +137,55 @@ void MoLAttentionLayer::forwarding(RunLayerContext &context, bool training) {
-  Tensor beta = beta_hat.apply(static_cast<float (*)(float)>(&std::exp));
+  kappa_src.apply_i(&expf);
+  beta_src.apply_i(&expf);
+  Tensor kappa = kappa_src;
+  Tensor beta = beta_src;
```
out of curiosity, wonder why we need this copy?

kparichay replied:
This is just for name-keeping and understanding. This will just be a shallow copy.

zhoonit replied:
Cool

zhoonit commented on nntrainer/layers/mol_attention_layer.cpp:
```diff
@@ -72,34 +137,55 @@ void MoLAttentionLayer::forwarding(RunLayerContext &context, bool training) {
 
-  Tensor mlp_out = query.dot(mlp_w);
+  /** reset helper state */
+  helper_exec = false;
```
what does this do?

kparichay replied:
Ensures that `calcGradient()` and `calcDerivative()` both dont call `calcDerivativeHelper()` twice.

zhoonit replied:
I see, thanks

zhoonit commented on nntrainer/layers/layer_context.cpp:
```diff
@@ -13,6 +13,7 @@
 #include "nntrainer_error.h"
 #include <functional>
 
+#include <iostream>
```
unnecessary header?

kparichay replied:
will remove.

zhoonit commented on nntrainer/layers/mol_attention_layer.cpp:
```diff
@@ -116,19 +202,159 @@ void MoLAttentionLayer::forwarding(RunLayerContext &context, bool training) {
+  dscores.reshape(TensorDim({scores.batch(), 1, scores.width(), 1}));
+
+  Tensor dprob_scaled = Tensor(TensorDim({batch, 1, value.height(), mol_k}));
+  dprob_scaled.setZero();
+  dprob_scaled.add_i(dscores);
```
Throwing some list for double checking
1. scores.batch() == batch ?
2. scores.width() == values.height()?

kparichay replied:
1. Skipped the batch size check as batch is assumed to be same across all inputs.
2. scores.width() is manually set to values.height() :) - https://github.com/nnstreamer/nntrainer/blob/775d4418ad208bcd9ba2a76d6c67f8c685a1c97e/nntrainer/layers/mol_attention_layer.cpp#L108

But the dimensions checks were completely missed in finalizing. Added dimensionality checks for values and query.

zhoonit replied:
👍

zhoonit commented on nntrainer/layers/mol_attention_layer.cpp:
```diff
@@ -116,19 +202,159 @@ void MoLAttentionLayer::forwarding(RunLayerContext &context, bool training) {
+  sigmoid.run_prime_fn(prob_right, du_neg_div, dprob_right);
+  Tensor du_neg_m = du_neg_div.divide(beta_eps);
+  Tensor dm_neg = du_neg_m.multiply(-1).sum(2);
+  Tensor dbeta_eps_neg = du_neg_m.multiply(u_neg_div).multiply(-1).sum(2);
```
considering) u_neg_div = u_neg_m/beta_eps

dbeta_eps = -d_u_neg_div * u_neg_m / (beta_eps ^ 2) ?

kparichay replied:
your formula is right. To reduce the amount of computation, consider the below subsitutions and simplified result used in the code:
```
du_neg_m = d_u_neg_div/beta_eps // given
u_neg_div = u_neg_m/beta_eps // given
dbeta_eps/du_neg_div = -d_u_neg_div * u_neg_m / (beta_eps ^ 2) = (d_u_neg_div/beta_eps) * (u_neg_m/beta_eps) * -1 = du_neg_m * u_neg_div
```

zhoonit replied:
That's clever. Thanks for the explanation :)

zhoonit commented on nntrainer/layers/mol_attention_layer.cpp:
```diff
@@ -116,19 +202,159 @@ void MoLAttentionLayer::forwarding(RunLayerContext &context, bool training) {
+  sigmoid.run_prime_fn(prob_left, du_pos_div, dprob_left);
+  Tensor du_pos_m = du_pos_div.divide(beta_eps);
+  Tensor dm_pos = du_pos_m.multiply(-1).sum(2);
+  Tensor dbeta_eps_pos = du_pos_m.multiply(u_pos_div).multiply(-1).sum(2);
```
same here??

kparichay replied:
same as above.

zhoonit commented on nntrainer/layers/mol_attention_layer.cpp:
```diff
@@ -116,19 +202,159 @@ void MoLAttentionLayer::forwarding(RunLayerContext &context, bool training) {
+  Tensor &dfc_proj_out =
+    context.getTensor(wt_idx[AttentionParams::fc_proj_out]);
+  Tensor &scores = context.getTensor(wt_idx[AttentionParams::scores]);
+  Tensor &prob = context.getTensor(wt_idx[AttentionParams::prob]);
+  Tensor &prob_left = context.getTensor(wt_idx[AttentionParams::prob_left]);
+  Tensor &prob_right = context.getTensor(wt_idx[AttentionParams::prob_right]);
+  Tensor &u_neg_div = context.getTensor(wt_idx[AttentionParams::u_neg_div]);
```
how about make this trainable tensor and use TensorGrad?

kparichay replied:
A lot of tensors are locally allocated in calcDerivativeHelper(). I have a todo to optimize for it and will use this suggestion when optimizing.

zhoonit commented on nntrainer/layers/mol_attention_layer.cpp:
```diff
@@ -137,7 +363,21 @@ void MoLAttentionLayer::setProperty(const std::vector<std::string> &values) {
+void MoLAttentionLayer::exportTo(Exporter &exporter,
+                                 const ExportMethods &method) const {
+  AttentionLayer::exportTo(exporter, method);
+  LayerImpl::exportTo(exporter, method);
```
setProperty will need this as well

kparichay replied:
added.

zhoonit commented on test/unittest/models/unittest_models.cpp:
```diff
@@ -40,11 +40,34 @@ IniWrapper reduce_mean_last("reduce_mean_last",
+  nn->setProperty({"batch_size=3"});
+
+  auto outer_graph = makeGraph({
+    {"input", {"name=in3", "input_shape=1:1:5"}},
+    {"input", {"name=in2", "input_shape=1:4:6"}},
+    {"input", {"name=in1", "input_shape=1:1:6"}},
```
Could we change this to fc to check if derivatives are correct?

kparichay replied:
I will update the #1758 unit test to include fc layer for derivatives.

zhoonit commented on test/input_gen/genModelTests_v2.py:
```diff
@@ -32,4 +69,10 @@ def forward(self, inputs, labels):
         name="reduce_mean_last",
     )
 
-    # inspect_file("lstm_single.nnmodelgolden")
```
Could you leave this at the last line for the reference?

kparichay replied:
done.

zhoonit commented on test/input_gen/genModelsRecurrent_v2.py:
```diff
@@ -190,6 +190,5 @@ def forward(self, inputs, labels):
         input_dims=[(3, 2)],
         label_dims=[(3, 2, 2)],
         name="grucell_stacked",
-    )
```
Maybe deleted by mistake?

kparichay replied:
yes, fixed.

zhoonit reviewed (approved):
LGTM. Amazing job done :)

jijoongmoon reviewed (approved):
LGTM
